### DISCUSSION TITLE
Resolving nested Target reference fails in partial reactor build

### DISCUSSION THREAD

**Discussion by vaclavHala** [eclipse-tycho/tycho#1209] [Q&A] [answered]
Hello,
we want to be able to run tests both with locally built bundles and with a materialized product.
To achieve this we have three target files:
* prod - contains bundles we ship in the product
* test - contains bundles used during testing such as junit
* all - contains reference to the two above targets via `<location type="Target" uri="file:${project_loc:prod.target}/prod.target"/>` and similar for `test.target`

In development tests (tycho surefire `default` runtime) we use the `all` target so that locally built bundles of the `prod` target can be used. (this target is also loaded as active target platform in eclipse)
In product tests (tycho surefire `p2Installed` runtime) we use just the `test` target in combination with `-Dtycho.localArtifacts=ignore` to be sure the bundles under test come from the installed product we are testing.

We have several test bundles with one parent which contains the setup described above in two profiles + all the test bundles and target files (each bundled individually as maven `eclipse-target` artifact) referenced as maven `module`.

Now when we run all the tests via the parent pom everything works great for both devel and product modes.
Problem is when we try to run just one of the test bundles in devel mode by moving into its directory and running plain `mvn clean verify` on that - the referred targets (`test` and `prod`) can't be found because they are not in the maven reactor.

* I tried running the single test bundle with addition of `-am path/to/prod.target` but that doesn't help.
* I also tried running the single test bundle by running the parent pom with addition of `-pl my_selected_test` but that doesn't work either
* what surprises me is that resolution of the `all.target` seems to pass but `prod.target` fails since the all technically isn't in the reactor either (but is selected by `target-configuration-plugin` so I guess there has to be some special handling for that)

I'm running tycho 2.7.4

I'm not sure if this is working as intended or it is a bug, any advice much appreciated

**Comment by laeubi** [comment]
to use `-am` / `-pl` you need to enable the tycho build extension:

https://github.com/eclipse/tycho/blob/master/RELEASE_NOTES.md#tycho-pomless-will-become-a-tycho-core-extension

**Comment by vaclavHala** [comment]
I should have said that all the artifacts have proper pom.xml so I'm not using pomless feature at all

**Comment by laeubi** [comment]
As the realse note suggest, we do not differentiate here anymore, there is only a `tycho-build` extension that also supports pomless builds but `-am` / `-pl` as well ...

**Comment by vaclavHala** [comment]
Hm I tried adding it but it made no difference. Maybe I misunderstood how it should be done - the project looks something like this
```
- tests
    - pom.xml # modules=[testsA, testsB,..releng/all.target,..releng/prod.target,..releng/test.target]
    - .mvn
        - extensions.xml # I added the extension you suggested here
    - testsA # I'm trying to run just this without running other test bundles
        - pom.xml # has .. as parent
    - testsB
        - pom.xml # has .. as parent
- releng
    - all.target
        - pom.xml
        - all.target
    - prod.target
        - pom.xml
        - prod.target
    - test.target
```

**Comment by laeubi** [comment, marked as answer]
You need to put it in the "root" and call e.g. `mvn -pl tests/testsA -am`

**Comment by vaclavHala** [comment]
Turns out the `-p` didn't work for me because I was calling it with artifact IDs but it expects exact string as present in `<module>`.
So I was calling `mvn -pl testsA,prod.target,test.target,all.target` which didn't find the targets. 
When called like this `mvn -pl testsA,../releng/prod.target,../releng/test.target,../releng/all.target` it works as expected.

I though it should be possible to use exact strings as printed by `Reactor Summary` to log but I guess this is just maven behavior.


As for why the resolution of `all` works but not `prod` and `test` I suppose that `all` is found via the maven GAV coordinates without looking for it as "project" but because the `prod` and `test` and referenced by `project_loc` so they can't be found.

Thanks for the quick help!

**Comment by laeubi** [comment]
> Turns out the `-p` didn't work for me because I was calling it with artifact IDs but it expects exact string as present in `<module>`.

If you want to use artifact ids, you need to use `-pl :artifactid` (note the collon `:` in front of it!)
